src/Passes/Source/Rules/Factory.cpp

```diff
@@ -502,6 +507,82 @@ namespace quantum
              replace_branch_positive});
     }
 
+    void RuleFactory::optimiseSelectQuantumOne()
+    {
+        auto replace_select_positive = [](Builder& builder, Value* val, Captures& cap, Replacements& replacements) {
+            auto result = cap["result"];
+            auto cond   = llvm::dyn_cast<llvm::Instruction>(cap["cond"]);
+            // Replacing result
+            auto orig_instr = llvm::dyn_cast<llvm::Instruction>(val);
+            if (orig_instr == nullptr)
+            {
+                return false;
+            }
+
+            auto                      module   = orig_instr->getModule();
+            auto                      function = module->getFunction("__quantum__qir__read_result");
+            std::vector<llvm::Value*> arguments;
+            arguments.push_back(result);
+
+            if (!function)
+            {
+                std::vector<llvm::Type*> types;
+                types.resize(arguments.size());
+                for (uint64_t i = 0; i < types.size(); ++i)
+                {
+                    types[i] = arguments[i]->getType();
+                }
+
+                auto return_type = llvm::Type::getInt1Ty(val->getContext());
+
+                llvm::FunctionType* fnc_type = llvm::FunctionType::get(return_type, types, false);
+                function                     = llvm::Function::Create(
+                    fnc_type, llvm::Function::ExternalLinkage, "__quantum__qir__read_result", module);
+            }
+
+            builder.SetInsertPoint(llvm::dyn_cast<llvm::Instruction>(val));
+            auto new_call = builder.CreateCall(function, arguments);
+            if (cond == nullptr)
+            {
+                return false;
+            }
```

Comment:
Nit: 
Feels like this `if` can be moved all the way up, and placed right after the line 514. In that case we can avoid calculating `new_call`, `function`, `arguments`, and may be even `orig_instr`.

If line 514 and this fragment are placed right after the line 512, then we can avoid calculating `result` too.